src/mesh/boundary_standard.cxx

```diff
@@ -318,178 +325,169 @@ void BoundaryDirichlet::apply(Field3D &f,BoutReal t) {
-	  // x norm is located half way between first grid cell and guard cell.  
-	  // y norm is shifted by half a grid point because it is staggered. 
-	  BoutReal xnorm = 0.5*(   mesh->GlobalX(bndry->x) + mesh->GlobalX(bndry->x - bndry->bx) ); 
-	  BoutReal ynorm = 0.5*(   mesh->GlobalY(bndry->y) + mesh->GlobalY(bndry->y - 1) );
-					      
-	  for(int zk=0;zk<mesh->LocalNz;zk++) {
-	    if(fg)
-	      val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*zk/(mesh->LocalNz), t);
-	
-	    f(bndry->x,bndry->y,zk) = 2*val - f(bndry->x-bndry->bx, bndry->y-bndry->by, zk);
-						
-	    // Need to set second guard cell, as may be used for interpolation or upwinding derivatives
-	    for(int i=1;i<bndry->width;i++) {
-	      int xi = bndry->x + i*bndry->bx;
-	      int yi = bndry->y ;
-						
-	      f(xi, yi, zk) = 2*f(xi - bndry->bx, yi , zk) - f(xi - 2*bndry->bx, yi, zk);	
-	    }	
-	  }
-	}			
-      }
-    }
-  }
-  else {
-    // Standard (non-staggered) case
+    }
+    if(bndry->by < 0){
+      // Lower y boundary. Set one point inwards
+      for(; !bndry->isDone(); bndry->next1d()) {
+
+        BoutReal xnorm = localmesh->GlobalX(bndry->x);
+        BoutReal ynorm = 0.5*(   localmesh->GlobalY(bndry->y) + localmesh->GlobalY(bndry->y - bndry->by) );
+
+        for(int zk=0;zk<localmesh->LocalNz;zk++) {
+          if(fg){
+            val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*zk/(localmesh->LocalNz), t);
+          }
+          f(bndry->x,bndry->y - bndry->by, zk) = val;
+
+          // Need to set second guard cell, as may be used for interpolation or upwinding derivatives
+          for(int i=0;i<bndry->width;i++) {
+            int xi = bndry->x ;
+            int yi = bndry->y + i*bndry->by;
+
+            f(xi, yi, zk) = 2*f(xi, yi - bndry->by, zk) - f(xi, yi - 2*bndry->by, zk);
+          }
+        }
+      }
+    }
+    if(bndry->bx != 0){
+      // x boundaries
+      for(; !bndry->isDone(); bndry->next1d()) {
+        // x norm is located half way between first grid cell and guard cell.
+        // y norm is shifted by half a grid point because it is staggered.
+        BoutReal xnorm = 0.5*(   localmesh->GlobalX(bndry->x) + localmesh->GlobalX(bndry->x - bndry->bx) );
+        BoutReal ynorm = 0.5*(   localmesh->GlobalY(bndry->y) + localmesh->GlobalY(bndry->y - 1) );
+
+        for(int zk=0;zk<localmesh->LocalNz;zk++) {
+          if(fg)
+            val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*zk/(localmesh->LocalNz), t);
+
+          f(bndry->x,bndry->y,zk) = 2*val - f(bndry->x-bndry->bx, bndry->y-bndry->by, zk);
+
+          // Need to set second guard cell, as may be used for interpolation or upwinding derivatives
+          for(int i=1;i<bndry->width;i++) {
+            int xi = bndry->x + i*bndry->bx;
+            int yi = bndry->y ;
+
+            f(xi, yi, zk) = 2*f(xi - bndry->bx, yi , zk) - f(xi - 2*bndry->bx, yi, zk);
+          }
+        }
+      }
+    }
+  } else {
+    // CELL_CENTRE or CELL_ZLOW
```

Comment:
Do we not care about the Z location in the `CELL_ZLOW` case for 3D fields?

Reply:
Going through all the boundary operators again will be painful... There's a lot of cut-and-paste code in boundary_standard.cxx, we could probably cut it down massively by having a single set of `BoundaryOp::apply()` methods implemented for the `BoundaryOp` class, and just having the single point stencils, one staggered and one unstaggered, for each derived class.

That would change quite a lot of code (on the other hand, I've touched most of the lines in boundary_standard.cxx already, at least for whitespace) so maybe you'd rather put off until 4.3. Also, since probably nobody is using CELL_ZLOW it doesn't matter too much if it stays broken for a little while longer (at least it's less broken now, constant value boundary conditions work).

I don't like leaving things broken though, so I'll try to make a fixed version, but I'll put it in a new PR.

Reply:
I think we're going to put this PR off till 4.3 -- there's a few questions about `Field`/`FieldData` and the classes that do and don't inherit from them that we really need to resolve, and that will involve some possibly major structural changes.

All the boundary stuff needs a thorough look through and rethink as well. Can we use `Region`, or a subclass instead of `BoundaryRegion`? Is there a way we can template the boundary conditions on simple kernels and stencils? @d7919 has been doing something like this for derivatives, so maybe we can extend that work to here as well.